### ISSUE TITLE
VS Code support

### ISSUE THREAD

**Issue by seoriy** [JosefPihrt/Roslynator#38]
It would great to have Roslynator as Visual Studio Code extension.

**Comment by JosefPihrt** [comment]
I would like to, but currently it is not possible.

See https://github.com/OmniSharp/omnisharp-vscode/issues/43

**Comment by JosefPihrt** [comment]
https://www.strathweb.com/2017/05/using-roslyn-refactorings-with-omnisharp-and-visual-studio-code/

**Comment by snielsson** [comment]
So is this still blocked? Or is it now possible to make an VSCode Roslynator Add-in?

**Comment by JosefPihrt** [comment]
I don't think it is possible yet, you can check that the issue mentioned above is still open.

**Comment by JosefPihrt** [comment]
[How to enable Roslynator for VS Code](https://github.com/JosefPihrt/Roslynator/blob/master/docs/RoslynatorForVisualStudioCode.md)

**Comment by colin-young** [comment]
@JosefPihrt I've tried following those instructions, but get:

`System.NotSupportedException: An attempt was made to load an assembly from a network location which would have caused the assembly to be sandboxed in previous versions of the .NET Framework. This release of the .NET Framework does not enable CAS policy by default [...]`

I fixed that by "unblocking" the assemblies in Windows (7). Now I'm getting:

`Expected to use Roslynator.CSharp.CodeFixes, Version=1.6.30.0, Culture=neutral, PublicKeyToken=94dac60bcac1903d in host services but the assembly cannot be loaded due to an exception: Could not load file or assembly 'Roslynator.Workspaces, Version=1.6.30.0, Culture=neutral, PublicKeyToken=8ca34578834dd224' or one of its dependencies. The system cannot find the file specified..
System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.`

Any suggestions?

**Comment by JosefPihrt** [comment]
Copying **Roslynator.Workspaces.dll** to your folder should solve the issue.

I have recently move code dependent on **Microsoft.CodeAnalysis.CSharp.Workspaces** to new assembly **Roslynator.Workspaces** but I have not update documentation yet.

**Comment by colin-young** [comment]
Thanks! One other thing to note is that I needed to restart VS Code to get it to load the DLLs. Simply restarting the Omnisharp extension was not sufficient. Maybe that's only the case when there are problems reading assemblies, and it's not a general problem if you manage to get your configuration right on your first try.

1. Extract DLLs
2. Unblock DLLs in Windows
3. Add/create settings in omnisharp.json
4. Start (or restart) VS Code

**Comment by JosefPihrt** [comment]
I'm glad it works!

I will update documentation accordingly.

**Comment by VahidN** [comment]
A PowerShell script to automate this process
```powershell
Write-Host "Download, unzip and enable Roslynator for Visual Studio Code"

$name = "josefpihrt.Roslynator2017"
$url = "https://marketplace.visualstudio.com/items?itemName=$name"
$currentDir = $PSScriptRoot
$file = "$currentDir\Roslynator.zip"

$pattern = "<script class=`"vss-extension`" defer=`"defer`" type=`"application\/json`">(.*?)<\/script>"
$regex = [regex]"(?m)$pattern"
Write-Host "Grab the home page of the $name."
$dom = (New-Object Net.WebClient).DownloadString($url); 
if($dom -and $dom -match $pattern) 
{
    $matches = $regex.Match($dom)
    $jsonText = $matches[0].Groups[1]

    $json = ConvertFrom-Json $jsonText

    $version = $Json.versions[0].version # Parse the json in the page for the latest version number
    $parts = $name.Split(".")
    $publisher = $parts[0]
    $package = $parts[1]

    # Assemble the url for the vsix package	
    $packageUrl = "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/$publisher/vsextensions/$package/$version/vspackage"	
    Write-Host "Download the vsix package: $packageUrl"
    (New-Object Net.WebClient).DownloadFile($packageUrl, $file)
	
    Write-Host "Using $currentDir as the current dir."
    Write-Host "Unzip $file."
    $shellApp = new-object -com shell.application 
    $zipFile = $shellApp.namespace($file) 
    $destination = $shellApp.namespace($currentDir) 
    $destination.Copyhere($zipFile.items(), 0x14)	# overwrite and be silent
	
    Write-Host "Delete VS specific files. Otherwise they will interfere with the MEF services inside OmniSharp."
    Remove-Item "$currentDir\Roslynator.VisualStudio.Core.dll","$currentDir\Roslynator.VisualStudio.dll", "$currentDir\Roslynator.VisualStudio.pkgdef"
	
    $omnisharpJsonFilePath = "$env:USERPROFILE\.omnisharp\omnisharp.json";	
    Write-Host "Create $omnisharpJsonFilePath file."
    $omnisharpJson = @" 
{{
  "RoslynExtensionsOptions": {{
    "LocationPaths": [
      "{0}"
    ]
  }}
}}
"@ -f $currentDir -Replace "\\","\\"
    $omnisharpJson | Out-File "$omnisharpJsonFilePath" -Confirm
	
    Write-Host "Done!"
}
else
{
    Write-Host "Failed to find the packageUrl!"
}
```

**Comment by domints** [comment]
Are these DLLs .NET Standard? I mean, would that work with VS Code on Mac?

**Comment by JosefPihrt** [comment]
Roslynator targets .NET Standard 1.3

**Comment by domints** [comment]
I'll give it a try then, and report the results :)

**Comment by divinebovine** [comment]
Roslyn analyzers and code fixes support was merged recently into [omnisharp-roslyn](https://github.com/OmniSharp/omnisharp-roslyn/pull/1076) so this should now be possible.

Until a new preference gets shipped and an updated stable release of omnisharp-roslyn, you'll need to use `"omnisharp.path": "latest"` in your settings.json and the following in your omnisharp.json:

```
{
	"RoslynExtensionsOptions": {
        "EnableAnalyzersSupport": true,
		"LocationPaths": [
			"C:\\lib\\roslynator"
		]
    }
}
```

I've tested this recently and was able to get both analyzer warnings/errors as well as codefixes.

**Comment by divinebovine** [comment]
Looks like the new preference is available in the latest omnisharp. This will remove the need for the "EnableAnalyzersSupport" setting in your omnisharp.json.

![image](https://user-images.githubusercontent.com/2818169/56430674-b00b8800-628c-11e9-9829-e0cab87ab453.png)

**Comment by divinebovine** [comment]
Can we remove the blocked label at this point?

**Comment by JosefPihrt** [comment]
Does anybody know how to pack roslyn libraries as VS Code extension?

**Comment by JosefPihrt** [comment]
Roslynator for Visual Studio Code is here! https://marketplace.visualstudio.com/items?itemName=josefpihrt-vscode.roslynator

**Comment by sbmw** [comment]
This is fantastic news!

One thing.  Currently user-level rules can be set by creating files at %LOCALAPPDATA%\JosefPihrt\Roslynator\.  This isn't possible for people running VS Code on non-Windows environments.

I think you could just update roslyn/common/roslynator.config within the extension to include a couple of extra imports like:

```
<Include Path="~/roslynator.config" />
...
<RuleSet Path="~/roslynator.ruleset" />
```

and users would be able to create a config and ruleset in their home directory?

**Comment by JosefPihrt** [comment]
Is anybody able to test that following code snippet will work on non-Windows environment?
```cs
using System;
using System.IO;
using System.Diagnostics;

string localAppDataPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);

string path = @"%LOCALAPPDATA%\JosefPihrt\Roslynator\VisualStudioCode";

path = path.Replace("%LOCALAPPDATA%", localAppDataPath);

// normalize directory separator
path = Path.GetFullPath(path);

// LINUX
Debug.Assert(path == "/home/foo/.local/share/JosefPihrt/Roslynator/VisualStudioCode");

// OSX
Debug.Assert(path == "/Users/foo/.local/share/JosefPihrt/Roslynator/VisualStudioCode");
```

Thanks.

**Comment by domints** [comment]
When I replaced `/home/foo/` with my current user directory on Linux Assert failed because of wrong slashes. 
Otherwise path looks fine (I added Console.WriteLine) :) 

My Linux is some kind of Debian (9 I think) and my dotnet version is 2.2.300

**Comment by sbmw** [comment]
@JosefPihrt yes, your way is definitely better than adding more paths.

As domints says, the directory normalizing isn't working.

You could do something like

```
if (Environment.OSVersion.Platform != PlatformID.Win32NT)
{
    path = path.Replace(@"\", "/");
}
```

Or else I think just using forward slashes as directory separators will work across all platforms?

**Comment by divinebovine** [comment]
You could use Path.PathSeparator for platform aware separators

**Comment by domints** [comment]
@sbmw that's not the good way :) I believe System.IO.Path has field specifying the correct path delimiter. It's `DirectorySeparatorChar`. But as `AltDirectorySeparatorChar` specifies on Windows, Windows also understands `/` slash, so that might be a way to go. 

I think `GetFullPath` didn't fix slashes on Linux, because neither `DirectorySeparatorChar` nor `AltDirectorySeparatorChar` specify `\` as proper separator on Linux.

@divinebovine `Path.PathSeparator` is meant to separate multiple paths, not directories in single path. It resolves to `;` on Windows and `:` on Linux.

**Comment by ADIX7** [comment]
There is Path.Combine to combine path parts.

**Comment by divinebovine** [comment]
@domints your right, my mistake

**Comment by domints** [comment]
@ADIX7

> There is Path.Combine to combine path parts.

yet it doesn't fix anything :)

**Comment by ADIX7** [comment]
Realy, I haven't read back enough posts... On windows doesn't / work? Or you can expect / and replace it with Path.DirectorySeparatorChar. 

According to this / should work on windows too:
https://docs.microsoft.com/en-us/dotnet/api/system.io.path.directoryseparatorchar?view=netframework-4.8

**Comment by JosefPihrt** [comment]
I think this should work:
```cs
string localAppDataPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);

string path = @"%LOCALAPPDATA%/JosefPihrt/Roslynator/VisualStudioCode";

path = path.Replace("%LOCALAPPDATA%", localAppDataPath);

// "/home/foo/.local/share/JosefPihrt/Roslynator/VisualStudioCode" on Linux with user "foo"
Console.WriteLine(path);
```
On Windows there will be mixed slashes but I have checked that it's fine.

**Comment by wocar** [comment]
how can i get roslynator to work with macos

**Comment by domints** [comment]
> how can i get roslynator to work with macos

Just install VS Code extension and modify one file according to readme on extension page

**Comment by wocar** [comment]
Can’t find it

**Comment by domints** [comment]
![image](https://user-images.githubusercontent.com/1096161/64360912-bfa52680-d00b-11e9-89d6-77dd568e21ce.png)
It is there ;)

**Comment by wocar** [comment]
Ok the assemblies seem to load but still doesnt work :(  

Can someone else confirm that works for mac

**Comment by domints** [comment]
@wocar 
![obraz](https://user-images.githubusercontent.com/1096161/64456307-361b5480-d0f0-11e9-8490-a33cd28c3b65.png)

**Comment by JosefPihrt** [comment]
I'm going to close this issue. Please file a new issue for anything related to VS Code. Thanks.
